File: 7. Valid Anagram/Hash-Map/valid-anagram-hash-map.ts
const isAnagram = (s: string, t: string): boolean => {
  if (s.length != t.length) {
    return false;
  }

  const sCount: Record<string, number> = {};
  const tCount: Record<string, number> = {};

  for (let i = 0; i < s.length; i++) {
    sCount[s[i]] = (sCount[s[i]] || 0) + 1;
    tCount[t[i]] = (tCount[t[i]] || 0) + 1;
  }

  for (const key in sCount) {
    if (sCount[key] !== tCount[key]) {
      return false;
    }
  }
  return true;
}

const s = 'racecar', t = 'carrace';

console.log(`String ${s} is an anagram of ${t}: ${isAnagram(s, t) ? 'Yes' : 'No'}`);

File: 7. Valid Anagram/Hash-Map/valid-anagram-hash-map.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution {
  public:
    bool isAnagram(string s, string t) {
      if (s.size() != t.size()) {
        return false;
      }
      unordered_map<char, int> sCount;
      unordered_map<char, int> tCount;

      for (int i = 0; i < s.size(); ++i) {
        sCount[s[i]]++;
        tCount[t[i]]++;
      }

      return sCount == tCount;
    }
};

int main() {
  Solution s1;
  string s = "racecar", t = "carrace";

  cout << "'" << s << "' is an anagram of '" << t << "' ";
  cout << ((s1.isAnagram(s, t)) ? "Yes" : "No") << endl;
  return 0;
}